File: Easy/257_BinaryTreePaths.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {string[]}
 */
const binaryTreePaths = (root) => {
    let result = [];
    addPaths(root, result, "" + root.val);
    return result;
};

const addPaths = (node, result, path) => {
    if (!node.left && !node.right) {
        result.push(path);
        return;
    }

    if (node.left) addPaths(node.left, result, path + "->" + node.left.val);
    if (node.right) addPaths(node.right, result, path + "->" + node.right.val);
}

File: Easy/257_binary-tree-paths.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public List<String> binaryTreePaths(TreeNode root) {
        List<String> result = new ArrayList<>();

        // if (root == null) {
        // return result;
        // }

        List<Integer> list = new ArrayList<>();
        list.add(root.val);
        helper(result, list, root);

        return result;
    }

    private void helper(List<String> result, List<Integer> list, TreeNode root) {
        // when should we terminate recursion
        if (root == null) {
            return;
        }

        // && when should we add the unique solution to result
        if (root.left == null && root.right == null) {
            StringBuffer sb = new StringBuffer();
            for (int i = 0; i < list.size() - 1; i++) {
                sb.append(list.get(i)).append("->");
            }
            sb.append(list.get(list.size() - 1));
            result.add(sb.toString());
            return;
        }

        if (root.left != null) {
            list.add(root.left.val);
            helper(result, list, root.left);
            list.remove(list.size() - 1);
        }

        if (root.right != null) {
            list.add(root.right.val);
            helper(result, list, root.right);
            list.remove(list.size() - 1);
        }
    }
}